### PR TITLE
Compare applications helm charts before running helm upgrade

### LINKED ISSUE #12095: Reduce number of Helm upgrades in application-installation-controller
### Description of the feature you would like to add / User story

The application-installation-controller currently runs a full helm upgrade every time it reconciles an application. See https://github.com/kubermatic/kubermatic/issues/12078#issuecomment-1482709163 and discussion following it. Since Helm is very imperative (you tell it to run an upgrade with `helm upgrade`) but our controllers should work declaratively, I believe it would be better if the controller did not run an upgrade on every reconcile but only when it's actually needed.

### Solution details

- application-installation-controller only upgrades (i.e. increments revision) of Helm chart when an upgrade is needed.
- An upgrade is only needed when:
  - The chart version changes.
  - The values change.
  - A drift between chart and actually deployed resources is detected. This could happen because a resource got deleted. Check if re-using some logic found in https://github.com/databus23/helm-diff is feasible.

### Alternative approaches

Leave the current implementation in place, which increments the revision every hour.

### Use cases

This will decrease Helm operations and would have prevented #12078.

### Additional information

```[tasklist]
### Other applications imporvements
- [ ] https://github.com/kubermatic/kubermatic/issues/12821
- [ ] https://github.com/kubermatic/kubermatic/issues/12820
- [ ] https://github.com/kubermatic/kubermatic/issues/12846
```

### PR DESCRIPTION
**Pull request by cnvergence**
**What this PR does / why we need it**:

First step towards comparable upgrades, track changes to Helm chart appVersion, added values, and rendered manifests. Tracking resources on the cluster will be a follow-up to this one.

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubermatic/kubermatic/issues/12095

**What type of PR is this?**

/kind feature

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
Reduce number of Helm upgrades in application-installation-controller by tracking changes to Helm chart version, values and templated manifests.
```

**Documentation**:

```documentation
NONE
```

### CODE REVIEW

File: pkg/applications/helmclient/client.go [hunk 1]
```diff
@@ -241,10 +246,86 @@ func (h HelmClient) DownloadChart(url string, chartName string, version string,
 // Otherwise it upgrades the chart.
 // charLoc is the path to the chart archive (e.g. /tmp/foo/apache-1.0.0.tgz) or folder containing the chart (e.g. /tmp/mychart/apache).
 func (h HelmClient) InstallOrUpgrade(chartLoc string, releaseName string, values map[string]interface{}, deployOpts DeployOpts, auth AuthSettings) (*release.Release, error) {
-	if _, err := h.actionConfig.Releases.Last(releaseName); err != nil {
+	currentRelease, err := h.actionConfig.Releases.Last(releaseName)
+	if err != nil {
+		h.logger.Debugw("Installing helm release", "release", releaseName)
 		return h.Install(chartLoc, releaseName, values, deployOpts, auth)
```
Comment (embik) [inline pkg/applications/helmclient/client.go]: Just checking in here ... is this the best way to check if a release exists? Is there any way to distinguish errors (if `.Releases.Last` surfaces errors like "no connection")?

File: pkg/applications/helmclient/client.go [hunk 2]
```diff
@@ -241,10 +246,86 @@ func (h HelmClient) DownloadChart(url string, chartName string, version string,
...
+	if err != nil {
+		return false, err
+	}
+
+	newManifests, err := h.renderManifests(chart, newValues)
+	if err != nil {
+		return false, err
+	}
+
+	return !reflect.DeepEqual(chart.AppVersion(), currentRelease.Chart.AppVersion()) ||
```
Comment (embik) [inline pkg/applications/helmclient/client.go]: Do we need a `reflect.DeepEqual` for comparing two strings here?

**Review by embik — COMMENTED** [review]
Two smaller questions.

File: pkg/applications/helmclient/client.go (hunk 2 above)
Comment (cnvergence) [reply pkg/applications/helmclient/client.go]: ah, I think I left it here, thanks for catching

File: pkg/applications/helmclient/client.go (hunk 1 above)
Comment (cnvergence) [reply pkg/applications/helmclient/client.go]: was thinking about this improvement as well

**Review by embik — APPROVED** [review]
/approve

Finally! ❤️

**Review by SimonTheLeg — APPROVED** [review]
I think it makes sense to give this a whirl on dev and see how it behaves.
I am a bit concerned and suspect there are some performance downsides because now we render all the manifest, stuff them into one big string and then run a DeepEqual on it.
But let's see how it goes.

**Comment by cnvergence** [comment]
> I think it makes sense to give this a whirl on dev and see how it behaves. I am a bit concerned and suspect there are some performance downsides because now we render all the manifest, stuff them into one big string and then run a DeepEqual on it. But let's see how it goes.

We can leave out manifest rendering for now, until we want to progress with checking deployed resources on the cluster, either way I agree with you

**Pull request merged**